### ISSUE TITLE
Assign to __schema__ instead of Schema; load and dump helpers?

### ISSUE THREAD

**Issue by sloria** [lovasoa/marshmallow_dataclass#30]
It may make sense to assign the generated schema to `__schema__` instead of `Schema` for consistency with `dataclasses.dataclass`, which only assigns dunder methods.

```python
@dataclass
class Artist:
    name: str

Artist.__schema__  # => Schema class
```

To improve the DX, you could add `dump` and `load` helper methods.

```python
import datetime as dt

from marshmallow_dataclass import dataclass, load

@dataclass
class Album:
    title: str
    release_date: dt.date

album = load(Album, {"title": "Hunky Dory", "release_date": "1971-12-17"})
```

Schema contructor arguments can be passed as keyword arguments.

```python
albums = load(Album, input_data, many=True)
```

This could even open the door to conveniences like

```python
albums = load(List[Album], input_data)  # calls Album.__schema__(many=True).load(input_data)
```

What do you think?

**Comment by lovasoa** [comment]
The goal of this library is to generate a Schema, which can then be used for more than just serialization and deserialization, so I'd rather not hide the Schema itself. The capital S in Schema indicates that it is a class.

**Comment by lovasoa** [comment]
I don't see enough advantages here to justify breaking compatibility.

**Comment by sloria** [comment]
I think that hiding the schema makes it more clear that marshmallow_dataclasses' dataclasses behave just like normal dataclasses. By adding a public member `.Schema`, it makes it appear that the decorator may be modifying the behavior. I had to look at the code to see that it was only adding the one member.

In my application, I ended up using `class_schema` directly with a plain dataclass to make it clear what was going on.

Your call on this--I don't have a strong opinion about it. For my usage, using `class_schema` works just fine.

**Comment by lovasoa** [comment]
> By adding a public member .Schema, it makes it appear that the decorator may be modifying the behavior.
What were you thinking it was changing ? If this is confusing, then we should definitely improve the README and the documentation. Do you have any suggestions about how we could clear that misunderstanding in the documentation ?

**Comment by sloria** [comment]
Perhaps `class_schema` should be documented first. I'd revise the quickstart code snippet to something like

```python
from dataclasses import dataclass, field
from typing import List, Optional

import marshmallow
import marshmallow_dataclass


@dataclass
class Building:
    # field metadata is used to instantiate the marshmallow field
    height: float = field(metadata={"validate": marshmallow.validate.Range(min=0)})
    name: str = field(default="anonymous")


@dataclass
class City:
    name: Optional[str]
    buildings: List[Building] = field(default_factory=list)


CitySchema = marshmallow_dataclass.class_schema(City)

# Deserialize dict to an object
city: City = CitySchema().load(
    {"name": "Paris", "buildings": [{"name": "Eiffel Tower", "height": 324}]}
)

# Serializing object to a dict
city_dict: dict = CitySchema().dump(city)
```

You could explain in a section after that that `marshamallow_dataclasses.dataclass` is just a shortcut that runs `class_schema` and assigns the result to `.Schema`.

I'd be happy to send a PR revising the docs along those lines if you'd like. I'll probably wait after https://github.com/lovasoa/marshmallow_dataclass/pull/31 though to prevent conflicts.

**Comment by lovasoa** [comment]
OK !

**Comment by sloria** [comment]
Another possible benefit of the OP is that we could get rid of the `post_load` method on the `BaseSchema`. The schema would just deserialize to a dict that would be used as the kwargs to the dataclass's ctor.

With the `post_load` method removed, the base schema class could be exposed publicly, which opens the door to an API like:

```python
from dataclasses import dataclass
from marshmallow_dataclasses import Schema

@dataclass
class Artist:
    id: str

ArtistSchema = Schema.from_dataclass(Artist)
```

That would obviate the `base_class` argument to `class_schema` etc because you could just extend the base schema

```python
import marshmallow_dataclass

class BaseSchema(marshmallow_dataclass.Schema):
    def on_bind_field(self, field_name, field_obj):
        field_obj.data_key = (field_obj.data_key or field_name).upper()

UppercaseArtistSchema = BaseSchema.from_dataclass(Artist)
```

Not saying we should do all of this at once, but it could be a nice cleanup. The OP is at least an iterative step towards that.

**Comment by sloria** [comment]
Actually, the more I think about this, the more I doubt the usefulness of my OP and of the `dataclass` decorator in general.

The core API of this library is `class_schema`. Any other boilerplate-reducing abstractions probably belong in userland.

If I were using this library extensively, I could implement my OP in a few lines:

```python
import dataclasses
import marshmallow_dataclass


def schemaclass(_cls, base_schema=None, **kwargs):
    def wrap(cls):
        dc = dataclasses.dataclass(_cls, **kwargs)
        dc.__schema__ = marshmallow_dataclass.class_schema(
            _cls, base_schema=base_schema
        )

    return wrap if _cls is None else wrap(_cls)


def load(dc, data, **kwargs):
    return dc.__schema__(**kwargs).load(data)


def dump(dc, obj, **kwargs):
    return dc.__schema__(**kwargs).dump(obj)
```

But the disagreement in this issue is probably a sign that we shouldn't force this abstraction on other users.

So if I were to propose a Big Bad Breaking Change in this lib, it'd be to deprecate/remove `@dataclass` and just add it as a "recipe" in the documentation.

**Comment by lovasoa** [comment]
I don't think there is really a need to deprecate `@dataclass`. It is a small function, and it will not be a pain to maintain. We can just make the change you suggested in the README, making the first example about `class_schema`, and talking about `@dataclass` only below. I developed this library for my own needs initially, and it was all about reducing boilerplate. I found that `@dataclass` helped me maintain short, clean, and well-organized model definitions. I can completely understand why one wouldn't like it, though. Let's just make it clear in the README that the central part of this library is `class_schema`, and that we also provide a few niceties like `@add_schema` and `@dataclass`.
